consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,177 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows this node to identify itself to the peer as a validator and vice versa
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	errCh := make(chan error, 4)
+	isValidatorCh := make(chan bool, 1)
+
+	go func() {
+		validatorProofMessage, err := sb.generateValidatorProofMessage(peer)
+		errCh <- err
+		if err != nil {
+			return
+		}
+		msgBytes, err := validatorProofMessage.Payload()
+		errCh <- err
+		if err != nil {
+			return
+		}
+		errCh <- peer.Send(istanbulValidatorProofMsg, msgBytes)
+	}()
+	go func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		errCh <- err
+		isValidatorCh <- isValidator
+	}()
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	// We write to errCh four times unless if the timeout is triggered or an error isn't nil
+	for i := 0; i < 4; i++ {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return false, err
+			}
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		}
+	}
+	return <-isValidatorCh, nil
+}
+
+// generateValidatorProofMessage will create a message that contains a proof
+// to the peer during the handshake that this node is a validator, which is just
+// a versioned enode message.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node
+// and an empty message is created.
+// If this node is a proxy, it will use the most recent versioned enode message this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
```

Comment:
I'm hesitant to bring out the `shouldSendValidatorProof` call because then it'd either involve an awkward boolean `shouldSendProof` or it would mean removing this function and instead putting the logic in the call site.

The content of the message is really dependent on the peer and I wanted to separate the logic for deciding what content to send in its own function, following the pattern of the other `generate` functions in `announce.go`. Is there a way I could beef up the comments to make this more clear?

Reply:
Looking at the `generateValidatorProofMessage` again after your latest refactor, the logic seems to be slimmed down a bit and I would support moving this functions contents to the call site. I think it'll end up being pretty short, and easier to read.

Reply:
It would also remove the need to pass the second return value which amounts to `peer.PurposeIsSet(p2p.ValidatorPurpose)`

Reply:
I agree, good suggestion! fixed